sdmetrics/single_table/data_augmentation/base.py

```diff
@@ -0,0 +1,258 @@
+"""Base class for Efficacy metrics for single table datasets."""
+
+from copy import deepcopy
+
+import numpy as np
+import pandas as pd
+from sklearn.metrics import confusion_matrix, precision_recall_curve, precision_score, recall_score
+from sklearn.preprocessing import OrdinalEncoder
+from xgboost import XGBClassifier
+
+from sdmetrics.goal import Goal
+from sdmetrics.single_table.base import SingleTableMetric
+from sdmetrics.single_table.data_augmentation.utils import _validate_inputs
+
+METRIC_NAME_TO_METHOD = {'recall': recall_score, 'precision': precision_score}
+
+
+class BaseDataAugmentationMetric(SingleTableMetric):
+    """Base class for Data Augmentation metrics for single table datasets."""
+
+    name = None
+    metric_name = None
+    goal = Goal.MAXIMIZE
+    min_value = 0.0
+    max_value = 1.0
+
+    def _fit_preprocess(self, data, metadata):
+        """Fit the preprocessing steps before applying the data augmentation technique."""
+        self._discrete_columns = []
+        self._datetime_columns = []
+        for column, column_meta in metadata['columns'].items():
+            if (column_meta['sdtype'] in ['categorical', 'boolean']) and (
+                column != self.prediction_column_name
+            ):
+                self._discrete_columns.append(column)
+            elif column_meta['sdtype'] == 'datetime':
+                self._datetime_columns.append(column)
+
+        self._ordinal_encoder = OrdinalEncoder()
+        self._ordinal_encoder.fit(data[self._discrete_columns])
+
+    def _fit(
+        self, data, metadata, prediction_column_name, minority_class_label, classifier, fixed_value
+    ):
+        """Fit the data augmentation technique."""
+        self.prediction_column_name = prediction_column_name
+        self.minority_class_label = minority_class_label
+        self.fixed_value = fixed_value
+        # To assess the preicision efficacy, we have to fix the recall and reciprocally
+        self._metric_to_fix = 'recall' if self.metric_name == 'precision' else 'precision'
+        self._metric_method = METRIC_NAME_TO_METHOD[self._metric_to_fix]
+        self._classifier_name = classifier
+        self._classifier = XGBClassifier(enable_categorical=True)
+
+        self._fit_preprocess(data, metadata)
+
+    def _transform_preprocess(self, tables):
```

Comment:
minor: we could just call transform

Reply:
I find `_fit_preprocess` and `_transform_preprocess` clearer, and there is no confusion with the `fit` method of the ML model. The ML model does not have a transform method, but since we keep `_fit_preprocess`, it's consistent to name `_transform_preprocess` as well.

Reply:
The ML model was separated out though. I think _fit_preprocess confuses me as it makes me wonder if we are fitting on preprocessed data or fitting as part of the preprocess step

Reply:
@R-Palazzo

Reply:
@amontanez24 sorry I missed this, done in [fbc5d37](https://github.com/sdv-dev/SDMetrics/pull/714/commits/fbc5d372879c7313dc46a13543158e6a629a8110)